################################################################################
SHORT DESCRIPTION: 
################################################################################
Expose a spinlock on IO port 0x3800.

################################################################################
LONG DESCRIPTION: 
################################################################################
reading 0x3800 returns 0x0 if locked, 0xffff else.
writing 0x3800 with anything !0 will take the lock.

VGABIOS was found to require some locking logic for its int10h handler. Having
an IO port to behave like a lock makes it much more easier to manage in VGABIOS
16b code.

################################################################################
CHANGELOG 
################################################################################
Documented: Eric Chanudet, chanudete@ainfosec.com, 17/03/2015
Port to QEMU 1.4.0: Eric Chanudet, chanudete@ainfosec.com, 01/03/2015
Port to QEMU 2.6.2: Ross Philipson, philipsonr@ainfosec.com, 10/13/2016

################################################################################
REMOVAL 
################################################################################
This patch is required by VGABIOS: vga-spinlock.patch.

################################################################################
UPSTREAM PLAN 
################################################################################
This is an OpenXT work-around. There is no plan to upstream this patch.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################
Implemented by VGABIOS: vga-spinlock.patch.
Depended on by QEMU: vga-shadow-bda.patch.

################################################################################
PATCHES 
################################################################################
--- a/hw/display/vga-isa.c
+++ b/hw/display/vga-isa.c
@@ -40,6 +40,7 @@ typedef struct ISAVGAState {
     struct VGACommonState state;
     PortioList portio_vga;
     PortioList portio_vbe;
+    PortioList portio_oxt;
 } ISAVGAState;
 
 static void vga_isa_reset(DeviceState *dev)
@@ -56,18 +57,20 @@ static void vga_isa_realizefn(DeviceStat
     ISAVGAState *d = ISA_VGA(dev);
     VGACommonState *s = &d->state;
     MemoryRegion *vga_io_memory;
-    const MemoryRegionPortio *vga_ports, *vbe_ports;
+    const MemoryRegionPortio *vga_ports, *vbe_ports, *oxt_ports;
 
     s->global_vmstate = true;
     vga_common_init(s, OBJECT(dev));
     s->legacy_address_space = isa_address_space(isadev);
-    vga_io_memory = vga_init_io(s, OBJECT(dev), &vga_ports, &vbe_ports);
+    vga_io_memory = vga_init_io(s, OBJECT(dev), &vga_ports, &vbe_ports, &oxt_ports);
     isa_register_portio_list(isadev, &d->portio_vga,
                              0x3b0, vga_ports, s, "vga");
     if (vbe_ports) {
         isa_register_portio_list(isadev, &d->portio_vbe,
                                  0x1ce, vbe_ports, s, "vbe");
     }
+    isa_register_portio_list(isadev, &d->portio_oxt,
+                             VGA_OXT_BASE, oxt_ports, s, "oxt");
     memory_region_add_subregion_overlap(isa_address_space(isadev),
                                         0x000a0000,
                                         vga_io_memory, 1);
--- a/hw/display/vga.c
+++ b/hw/display/vga.c
@@ -1818,6 +1818,11 @@ static void vga_update_display(void *opa
 
     qemu_flush_coalesced_mmio_buffer();
 
+    if (__sync_lock_test_and_set(&s->locked, 1)) {
+        /* Try to take spinlock or bail */
+        return;
+    }
+
     if (surface_bits_per_pixel(surface) == 0) {
         /* nothing to do */
     } else {
@@ -1845,6 +1850,8 @@ static void vga_update_display(void *opa
             break;
         }
     }
+
+    __sync_lock_release(&s->locked);
 }
 
 /* force a full display refresh */
@@ -2267,6 +2274,7 @@ void vga_common_init(VGACommonState *s,
     s->get_offsets = vga_get_offsets;
     s->get_resolution = vga_get_resolution;
     s->hw_ops = &vga_ops;
+    s->locked = 0;
     switch (vga_retrace_method) {
     case VGA_RETRACE_DUMB:
         s->retrace = vga_dumb_retrace;
@@ -2310,15 +2318,45 @@ static const MemoryRegionPortio vbe_port
     PORTIO_END_OF_LIST(),
 };
 
+static uint32_t spinlock_ioport_read(void *opaque, uint32_t addr)
+{
+    VGACommonState *s = opaque;
+
+    (void) addr;
+    if (__sync_lock_test_and_set(&s->locked, 1)) {
+        /* Can't lock */
+        return 0;
+    } else {
+        /* Locked */
+        return 0xffff;
+    }
+}
+static void spinlock_ioport_write(void *opaque, uint32_t addr, uint32_t val)
+{
+    VGACommonState *s = opaque;
+
+    (void) addr;
+    if (val) {
+        __sync_lock_release(&s->locked);
+    }
+}
+
+static const MemoryRegionPortio oxt_portio_list[] = {
+    { 0, 1, 2, .read = spinlock_ioport_read, .write = spinlock_ioport_write }, /* 0x3800 */
+    PORTIO_END_OF_LIST(),
+};
+
 /* Used by both ISA and PCI */
 MemoryRegion *vga_init_io(VGACommonState *s, Object *obj,
                           const MemoryRegionPortio **vga_ports,
-                          const MemoryRegionPortio **vbe_ports)
+                          const MemoryRegionPortio **vbe_ports,
+                          const MemoryRegionPortio **oxt_ports)
 {
     MemoryRegion *vga_mem;
 
     *vga_ports = vga_portio_list;
     *vbe_ports = vbe_portio_list;
+    *oxt_ports = oxt_portio_list;
 
     vga_mem = g_malloc(sizeof(*vga_mem));
     memory_region_init_io(vga_mem, obj, &vga_mem_ops, s,
@@ -2332,7 +2370,7 @@ void vga_init(VGACommonState *s, Object
               MemoryRegion *address_space_io, bool init_vga_ports)
 {
     MemoryRegion *vga_io_memory;
-    const MemoryRegionPortio *vga_ports, *vbe_ports;
+    const MemoryRegionPortio *vga_ports, *vbe_ports, *oxt_ports;
 
     qemu_register_reset(vga_reset, s);
 
@@ -2340,7 +2378,7 @@ void vga_init(VGACommonState *s, Object
 
     s->legacy_address_space = address_space;
 
-    vga_io_memory = vga_init_io(s, obj, &vga_ports, &vbe_ports);
+    vga_io_memory = vga_init_io(s, obj, &vga_ports, &vbe_ports, &oxt_ports);
     memory_region_add_subregion_overlap(address_space,
                                         0x000a0000,
                                         vga_io_memory,
@@ -2355,6 +2393,8 @@ void vga_init(VGACommonState *s, Object
         portio_list_init(&s->vbe_port_list, obj, vbe_ports, s, "vbe");
         portio_list_add(&s->vbe_port_list, address_space_io, 0x1ce);
     }
+    portio_list_init(&s->oxt_port_list, obj, oxt_ports, s, "openxt");
+    portio_list_add(&s->oxt_port_list, address_space_io, VGA_OXT_BASE);
 }
 
 void vga_init_vbe(VGACommonState *s, Object *obj, MemoryRegion *system_memory)
--- a/hw/display/vga_int.h
+++ b/hw/display/vga_int.h
@@ -37,6 +37,9 @@
 #define CH_ATTR_SIZE (160 * 100)
 #define VGA_MAX_HEIGHT 2048
 
+#define VGA_OXT_BASE                   0x3800
+#define VGA_OXT_SPINLOCK               VGA_OXT_BASE
+
 struct vga_precise_retrace {
     int64_t ticks_per_char;
     int64_t total_chars;
@@ -100,6 +103,7 @@ typedef struct VGACommonState {
                         int *pheight);
     PortioList vga_port_list;
     PortioList vbe_port_list;
+    PortioList oxt_port_list;
     /* bochs vbe state */
     uint16_t vbe_index;
     uint16_t vbe_regs[VBE_DISPI_INDEX_NB];
@@ -149,6 +153,8 @@ typedef struct VGACommonState {
     vga_update_retrace_info_fn update_retrace_info;
     union vga_retrace retrace_info;
     uint8_t is_vbe_vmstate;
+    /* ioport spinlock */
+    int locked;
 } VGACommonState;
 
 static inline int c6_to_8(int v)
@@ -164,7 +170,8 @@ void vga_init(VGACommonState *s, Object
               MemoryRegion *address_space_io, bool init_vga_ports);
 MemoryRegion *vga_init_io(VGACommonState *s, Object *obj,
                           const MemoryRegionPortio **vga_ports,
-                          const MemoryRegionPortio **vbe_ports);
+                          const MemoryRegionPortio **vbe_ports,
+                          const MemoryRegionPortio **oxt_ports);
 void vga_common_reset(VGACommonState *s);
 
 void vga_dirty_log_start(VGACommonState *s);
